Source/UI/NewUI/BillingForm/View/Field/BillingFormTextFieldView.swift

```diff
@@ -0,0 +1,171 @@
+    private weak var delegate: BillingFormTextFieldViewDelegate?
+    
+    private(set) lazy var headerLabel: UILabel = {
+        let view = UILabel()
+        view.text = style.title?.text
+        view.font = style.title?.font
+        view.textColor = style.title?.textColor
+        view.numberOfLines = 0
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .white
+        return view
+    }()
+    
+    private(set) lazy var hintLabel: UILabel = {
+        let view = UILabel()
+        view.text = style.hint?.text
+        view.font = style.hint?.font
+        view.textColor = style.hint?.textColor
+        view.numberOfLines = 0
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .white
+        return view
+    }()
+    
+    private(set) lazy var textFieldContainer: UIView = {
+        let view = UIView()
+        view.layer.cornerRadius = 10.0
+        view.layer.borderWidth = 1.0
+        textField.autocorrectionType = .no
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.layer.borderColor = style.textfield.normalBorderColor.cgColor
+        view.backgroundColor = style.textfield.backgroundColor
+        textField.keyboardType = style.textfield.isSupprtingNumbericKeyboard ? .phonePad : .default
+        textField.textContentType = style.textfield.isSupprtingNumbericKeyboard ? .telephoneNumber : .name
+        return view
+    }()
+    
+    private(set) lazy var textField: UITextField = {
+        let view: UITextField = self.type == .phoneNumber ? BillingFormPhoneNumberText(type: type) : BillFormTextField(type: self.type)
+        view.text = style.textfield.text
+        view.font = style.textfield.font
+        view.placeholder = style.textfield.placeHolder
+        view.textColor = style.textfield.textColor
+        view.tintColor = style.textfield.tintColor
+        view.delegate = self
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .clear
+        return view
+    }()
+    
+    private(set) lazy var errorView: BillingFormTextFieldErrorView = {
+        let view = BillingFormTextFieldErrorView(style: style.error)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
+    }()
+    
+    init(type: BillingFormCellType, style: BillingFormTextFieldCellStyle, delegate: BillingFormTextFieldViewDelegate? = nil) {
+        self.style = style
+        self.type = type
+        super.init(frame: .zero)
+        self.delegate = delegate
+        self.setupViews()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+}
+
+extension BillingFormTextFieldView {
+
+    private func setupViews(){
+        backgroundColor = style.backgroundColor
+        setupHeaderLabel()
+        setupHintLabel()
+        setupTextFieldContainer()
+        setupTextField()
+        setupErrorView()
```

Comment:
I see that across the project this method of breaking down the setup of the UI was avoided. I think there is a good reason for it, as in your current approach, I can come and reorder your setup to a way of my linking. That will crash the app when I try to add a constraint between 2 views that don't share the same view hierarchy (ie:  pin top edge of errorView to bottom edge of text field, if text field is not already added as a subview).

I agree there is value in clearly separating and organising the setup, but that may be done better without using functions:
```
// Header

         addSubview(headerLabel)
         NSLayoutConstraint.activate([
             headerLabel.topAnchor.constraint(equalTo: topAnchor),
             headerLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
             headerLabel.trailingAnchor.constraint(equalTo: trailingAnchor)
         ])

// Hint Label

         addSubview(hintLabel)
         NSLayoutConstraint.activate([
             hintLabel.topAnchor.constraint(equalTo: headerLabel.safeBottomAnchor, constant: 2),
             hintLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
             hintLabel.trailingAnchor.constraint(equalTo: trailingAnchor)
         ])
.....
```

Reply:
question : @alex-ioja-yang-cko , But even if all the code is in one big function could anyone not change Header and label up and down to break the code ? That will be applicable to entire project. My understanding is such scenarios are better covered off using UI tests where it checks whats expected behaviour and order and breaks the UI test if necessary.

Reply:
Agree, someone can always do that, but moving a block of 6 lines requires more attention than an abstract method name. Also, the big block can make obvious the constraint between 2 views that were added at the same time, as they're in the same viewing range.


Alternatively, we can use an extension method `addSubviews`, in which we add all the equal level subviews at the same time, then break down the constraint creation safely as they will be in the same hierarchy at that point.

```
addSubviews([headerLabel, hintLabel])
setupHeaderLabel()
setupHintLabel()
```

I've only ever previously used UITests for Product Core Journey testing, so I'm not sure you could ever achieve 100% coverage once you start being concerned of the resources the UI test uses.
Snapshot testing may catch it though. Those are like unit tests with full UI lifecycle called.

Reply:
change naming to `setupViewsInOrder`